VISH

saadhamid96 opened the pull request.

saadhamid96 committed:
Add script that should run VISH. There is still buggy.

saadhamid96 committed:
Fix bug by passing compile=False flags into optimisation routine.

saadhamid96 committed:
Separate VISH components into their own module, and add warped version

saadhamid96 committed:
Fix map_to_sphere function

saadhamid96 committed:
Basic working version of mixed eigenfunction and pseudo-observation inducing variables

saadhamid96 committed:
Add mixed eigenfunction and pseudo-observation inducing variables

vdutor committed:
Remove dimensionality assert in chord kernel

saadhamid96 committed:
Add log-transform implementation

saadhamid96 committed:
Add example script to set q_mu and q_sqrt using optimal equations from Titsias (2009) for interdomain inducing variables

saadhamid96 committed:
Attempt to get lengthscale gradients working

saadhamid96 committed:
Modify implementation of lengthscale gradient slightly. Still not functional though

saadhamid96 committed:
Change gradient implementation for ChordMatern. Rather than wrapping _funk_hecke, wrap just the scipy integration routine

saadhamid96 committed:
Cache eigenvalues when making predictions

saadhamid96 committed:
Fix spherical_matern kernel so that posterior variance drops to approximately the likelihood variance at data points. ChordMatern is still broken in this regard.

vdutor commented on gspheres/kernels/chord_matern.py:
```diff
@@ -69,13 +120,33 @@ def _funk_hecke(shape_function: Callable[[float], float], n: int, dim: int) -> f
-
-    v = integrate.quad(integrand, -1.0, 1.0)[0]
-    return v * omega_d / C_1
+    @tf.custom_gradient
```
Could you write down in math what the gradients look like? That said, I don't think we need the gradients of funk hecke wrt the lengthscale and variance. For the lengthscale you can simply multiply the inputs `X` before evaluating the spherical harmonics and the variance can be accounted for by multiplying the eigenvalues with it.

saadhamid96 replied:
The gradient wrt the variance is just the funk hecke integral.
The gradient wrt the lengthscale is the funk hecke integral with the shape function replaced by the derivative of the shape function wrt to the lengthscale, i.e. v \int d/dl s(t, l) f(t) dt, where f(t) is the product of the Gegenbauer polynomial and the weighting function wrt which the Gegenbauers are orthogonal and v is the variance.

vdutor commented on gspheres/kernels/chord_matern.py:
```diff
@@ -23,35 +25,84 @@ def __init__(self, nu: int, dimension: int):
+            values.append(tf.reshape(v, shape=[-1]))
+        return tf.concat(values, axis=0)
+
+    def verify_eigenvalue_cache(self):
```
Why do you need this functionality. Minor: maybe a more descriptive name would be `clear_eigenvalue_cache_onchange`?

saadhamid96 replied:
To avoid having to recompute funk-hecke integrals. A call to the `eigenvalues` method is made every time the Kuu matrix is required, which means every time a prediction is required. If we didn't cache the eigenvalues then we will do a lot of redundant computation whilst optimising the acquisition function, since we need to recompute the posterior variance at every step.

vdutor commented on gspheres/kernels/chord_matern.py:
```diff
@@ -23,35 +25,84 @@ def __init__(self, nu: int, dimension: int):
 
     def K_diag(self, X: TensorType) -> tf.Tensor:
         """ Approximate the true kernel by an inner product between feature functions. """
+        # TODO: Refactor in terms of truncated Mercer decomposition.
```
why would you want to do that?

saadhamid96 replied:
So that the posterior variance at the data locations goes down to the likelihood variance. This is necessary for the acquisition function to work properly. The variance of the unwarped GP is combination of the posterior mean and posterior variance of the warped GP. If the posterior variance doesn't get sufficiently small, the acquisition function will keep selecting the peak of the mean function. (You can see this for yourself if you run the experiment but change line 38 of the config file `vish.yaml` from "spherical_matern" to "chord_matern".)

I appreciate that this issue is specific to using VISH for PI, so maybe you don't want it changed in this repo? If so, we could fork it, or just work in a different branch? We could also just keep using the SphericalMatern kernel, which is now working fine.

vdutor commented on gspheres/vish/__init__.py:
```diff
@@ -0,0 +1,92 @@
+from gspheres.vish.mixed_variables import MixedFeatures
+
+
+def map_to_sphere(X: np.ndarray, bias: Union[int, float]) -> np.ndarray:
```
minor: I would place this in a `utils.py`.

vdutor commented on gspheres/vish/eigenfunction_variables.py:
```diff
@@ -0,0 +1,200 @@
+import tensorflow as tf
```
minor: rename file to `covariances.py`?

vdutor commented on gspheres/vish/eigenfunction_variables.py:
```diff
@@ -0,0 +1,200 @@
+@cov.Kuu.register(SphericalHarmonicFeatures, gpflow.kernels.Kernel)
+def Kuu_sphericalmatern_sphericalharmonicfeatures(
+        inducing_variable,
+        kernel,
```
```suggestion
        inducing_variable: SphericalHarmonicFeatures,
        kernel: gpflow.kernels.Kernel,
```

vdutor commented on gspheres/vish/eigenfunction_variables.py:
```diff
@@ -0,0 +1,200 @@
+    ])
+    eigenvalues = chain(eigenvalues_per_level, num_harmonics_per_level)
+    return tf.linalg.LinearOperatorDiag(1 / eigenvalues)
+    # return tf.linalg.diag(1 / eigenvalues)
```
```suggestion
```

vdutor commented on gspheres/vish/eigenfunction_variables.py:
```diff
@@ -0,0 +1,200 @@
+def Kuf_sphericalmatern_sphericalharmonicfeatures(
+        inducing_variable,
+        kernel,
+        X
```
```suggestion
        inducing_variable: SphericalHarmonicFeatures,
        kernel: gpflow.kernels.Kernel,
        X: TensorLike
```

vdutor commented on gspheres/vish/eigenfunction_variables.py:
```diff
@@ -0,0 +1,200 @@
+    """
+    Covariance between spherical harmonic features and function values.
+    """
+
```
```suggestion
```

vdutor commented on gspheres/vish/eigenfunction_variables.py:
```diff
@@ -0,0 +1,200 @@
+    Covariance between spherical harmonic features and function values.
+    """
+
+    return tf.transpose(inducing_variable.spherical_harmonics(X))
```
Minor: best to add the shape of the return value.

vdutor commented on gspheres/vish/mixed_variables.py:
```diff
@@ -0,0 +1,66 @@
+import gpflow.kernels
```
Are you still using this? If not, best to remove the file from the PR.

saadhamid96 replied:
Not using it anymore; will remove.

vdutor commented on scripts/vish.py:
```diff
@@ -0,0 +1,94 @@
+import numpy as np
```
I would refactor some of this code and have it as an integration test too in `tests`.

saadhamid96 replied:
Will do

saadhamid96 committed:
Update gspheres/vish/eigenfunction_variables.py

Co-authored-by: Vincent Dutordoir <dutordoirv@gmail.com>

saadhamid96 committed:
Apply suggestions from code review

Co-authored-by: Vincent Dutordoir <dutordoirv@gmail.com>

saadhamid96 committed:
Update Kuf docstring with return shape.

saadhamid96 committed:
Rename eigenfunction_variables.py to covariances.py

vdutor committed:
VISH SVGP

vdutor committed:
VISH SVGP

vdutor committed:
Variance

vdutor committed:
nu float type

vdutor committed:
ARD lengthscales and variance

saadhamid96 committed:
Partial commit of edits made whilst on vincent/vish branch

saadhamid96 committed:
Add VishGPR model and comparison of TuncatedChordMatern and ChordMatern

saadhamid96 committed:
Work in updates from vincent/vish branch

saadhamid96 commented:
I added a model for SGPR in `vish/__init__.py`. You were right that simply setting the optimal variational parameters for the SVGP model wasn't a good idea because the matrix that needs to be inverted is not well conditioned. "Rotating" (more of a scaling since Kuu is diagonal) by the cholesky decomposition of Kuu does help in this regard.

I've also added a script to compare the posterior variances at data points between a TruncatedChordMatern kernel and a ChordMatern kernel. The difference between the two is that the prior variance is computed using a truncated mercer decomposition for the TruncatedChordMatern, and using the ordinary Matern equations for the ChordMatern. When the likelihood variance is allowed to be optimised the posterior variance at data points is below the likelihood variance for all kernels. (I've added a pytest that ensures this in `tests/test_vish_gpr.py`.) However, we need to keep the likelihood variance small if we want to use the VISH model with the uncertainty sampling acquisition function or we will keep making acquisitions at the same location. If we fix it to 0.01 or below (in practice we'll want it quite small, say 1e-4), then notice that the posterior variance at data points stays below the likelihood variance for the TruncatedChordMatern, but not for the ChordMatern.

vdutor committed:
Merge pull request #4 from vdutor/vincent/vish

VISH

saadhamid96 committed:
Minor formatting changes

saadhamid96 committed:
Merge branch 'master' into sgpr

saadhamid96 committed:
Minor fix
